internal/template-validator/tlsinfo/tlsinfo.go

```diff
@@ -174,31 +162,78 @@ func loadCertificates(directory string) (serverCrt *tls.Certificate, err error)
 	return &cert, nil
 }
 
-func (ti *TLSInfo) getCertificate() *tls.Certificate {
-	ti.certLock.Lock()
-	defer ti.certLock.Unlock()
-	return ti.cert
+func (ti *TLSInfo) updateTLSOptions(directory string) {
+	tlsOptions, err := loadTLSOptions(directory)
+
+	ti.lock.Lock()
+	defer ti.lock.Unlock()
+
+	if err != nil {
+		ti.minTLSVersion = 0
+		ti.cipherSuites = nil
+		ti.tlsOptionsError = err
+		logger.Log.Error(err, "failed to load TLS options",
+			"directory", directory)
+		return
+	}
+
+	if tlsOptions.IsEmpty() {
+		// Using default configuration
+		ti.minTLSVersion = 0
+		ti.cipherSuites = nil
+		ti.tlsOptionsError = nil
+		logger.Log.Info("TLS options are empty, using default configuration")
+		return
+	}
+
+	minVersion, err := tlsOptions.MinTLSVersionId()
+	if err != nil {
+		ti.minTLSVersion = 0
+		ti.cipherSuites = nil
+		ti.tlsOptionsError = fmt.Errorf("TLS Configuration broken, min version misconfigured: %w", err)
+		logger.Log.Error(ti.tlsOptionsError, "TLS options are not valid",
+			"directory", directory)
+		return
+	}
+
+	ti.minTLSVersion = minVersion
+	ti.cipherSuites = common.CipherIDs(tlsOptions.OpenSSLCipherNames, nil)
```

Comment:
What happens if this leaves us with an empty slice in case of user misconfiguration?

Reply:
Good point. If user passes unknown cipher names, they will be silently left out. That is how it is implemented in the library function `crypto.OpenSSLToIANACipherSuites()`. I will add code to return an error instead.

Reply:
Thinking about this, it may be desired behavior. Usually, on TCP connection, the server gives client a list of supported ciphers and the client chooses. Any unknown ciphers are simply ignored.
Maybe it is ok to use the same approach for configuration file. User could specify multiple ciphers, and it is not an error if the validator does not know some of them.

What do you think?

Reply:
I was thinking what happens if you configure the API server with an empty slice if using TLS 1.1/1.2? Will it fallback to default ciphers? I'm not sure we should swallow this error silently. It is misconfiguration of the API server, not a failure in negotiating a cipher during a handshake.

Reply:
If we configure the server ciphers with empty slice `[]` and not use TLS 1.3, then no ciphers will be enabled, so it will not accept any connection. If we configure it with `nil` slice, then it will use default ciphers.
This is IMHO a bad API design in the standard library. Usually in other places there is no functional difference between `[]` and `nil`.

Ok, I will change the code, so it fails with error if ciphers in config file are not known.

Reply:
I've implemented the error checking and noticed that OpenShift's TLS security profiles also contain ciphers that are not implemented in the TLS in golang and they are silently ignored.
For example, here is the TLS `Intermediate` profile:
https://github.com/openshift/api/blob/a2817b89f7e0989016967055a142d5e88bed18de/config/v1/types_tlssecurityprofile.go#L288-L303
```go
TLSProfileIntermediateType: {
	Ciphers: []string{
		"TLS_AES_128_GCM_SHA256",
		"TLS_AES_256_GCM_SHA384",
		"TLS_CHACHA20_POLY1305_SHA256",
		"ECDHE-ECDSA-AES128-GCM-SHA256",
		"ECDHE-RSA-AES128-GCM-SHA256",
		"ECDHE-ECDSA-AES256-GCM-SHA384",
		"ECDHE-RSA-AES256-GCM-SHA384",
		"ECDHE-ECDSA-CHACHA20-POLY1305",
		"ECDHE-RSA-CHACHA20-POLY1305",
		"DHE-RSA-AES128-GCM-SHA256",
		"DHE-RSA-AES256-GCM-SHA384",
	},
	MinTLSVersion: VersionTLS12,
},
```

It contains ciphers `DHE-RSA-AES128-GCM-SHA256` and `DHE-RSA-AES256-GCM-SHA384`, which are not available in the golang TLS library.
The code that does the translation between names and cipher IDs ignores them: https://github.com/openshift/library-go/blob/64780a4acea5ee3e565683ad96b1f4f0799f0bd2/pkg/crypto/crypto.go#L149-L180

So it is not trivial to check if a cipher name is valid. We don't want it to fail if users set one of the OpenShift's profiles, but they already contain unknown ciphers. So we would need to specifically allow them, even if they are not implemented.

What do you think?

Reply:
As long as an empty slice that is not nil does not result in default values being used, I'm fine with it.

But strange that these profiles ship ciphers that are not supported. Are they supported in downstream distributions of golang?

Reply:
The downstream TLS library is the same as upstream, there are no changes.